### PR TITLE
pub support get total replicas from annotations

### PR DESCRIPTION
**Pull request by zmberg**
Signed-off-by: liheng.zms <liheng.zms@alibaba-inc.com>
### Ⅰ. Describe what this PR does
At present, the PUB protection must be the Pod under the workload. But there may be no workload in some scenarios. For this scenario, pub supports obtaining total replicas via pub.annotations[kruise.io/pub-protect-total-replicas]

### CODE REVIEW

File: pkg/control/pubcontrol/pub_control.go [hunk 1]
```diff
@@ -92,14 +93,17 @@ func (c *commonControl) GetPodsForPub(pub *policyv1alpha1.PodUnavailableBudget)
...
-
 	matchedPods := make([]*corev1.Pod, 0, len(podList.Items))
 	for i := range podList.Items {
 		pod := &podList.Items[i]
 		if kubecontroller.IsPodActive(pod) {
 			matchedPods = append(matchedPods, pod)
 		}
 	}
+	if pub.Annotations[policyv1alpha1.PubProtectTotalReplicas] != "" {
+		expectedCount, _ := strconv.ParseInt(pub.Annotations[policyv1alpha1.PubProtectTotalReplicas], 10, 32)
```
Comment (furykerry) [inline pkg/control/pubcontrol/pub_control.go:104]: shall we check workload targetRef first, if targetRef is nil , then use the annotation?

File: pkg/controller/podunavailablebudget/pub_controller_test.go [hunk 2]
```diff
@@ -754,6 +758,64 @@ func TestPubReconcile(t *testing.T) {
...
+			},
+			getReplicaSet: func() *apps.ReplicaSet {
+				object := replicaSetDemo.DeepCopy()
+				object.Spec.Replicas = utilpointer.Int32Ptr(100)
+				return object
+			},
+			getPub: func() *policyv1alpha1.PodUnavailableBudget {
+				pub := pubDemo.DeepCopy()
+				pub.Annotations[policyv1alpha1.PubProtectTotalReplicas] = "50"
+				for i := 0; i < 10; i++ {
```
Comment (furykerry) [inline pkg/controller/podunavailablebudget/pub_controller_test.go:788]: remove targetRef and set selector before testing?

Comment (zmberg) [reply pkg/controller/podunavailablebudget/pub_controller_test.go:790]: add pod.OwnerReferences = nil before testing.

File: pkg/util/controllerfinder/controller_finder.go [hunk 3]
```diff
@@ -120,9 +120,10 @@ func (r *ControllerFinder) GetExpectedScaleForPods(pods []*corev1.Pod) (int32, e
 		workload, err := r.GetScaleAndSelectorForRef(ref.APIVersion, ref.Kind, pod.Namespace, ref.Name, ref.UID)
 		if err != nil && !errors.IsNotFound(err) {
 			return 0, err
-		}
-		if workload != nil && workload.Metadata.DeletionTimestamp.IsZero() {
+		} else if workload != nil && workload.Metadata.DeletionTimestamp.IsZero() {
 			controllerScale[workload.UID] = workload.Scale
+		} else if workload != nil && !workload.Metadata.DeletionTimestamp.IsZero() {
```
Comment (furykerry) [inline pkg/util/controllerfinder/controller_finder.go:125]: if workload is deleted, no controllerScale entry will exists, it seems it is equivalent as setting controllerScale[workload.UID] = 0

**Review by furykerry — COMMENTED** [review]
/lgtm

**Pull request merged**